org.eclipse.xtext.ui.tests/tests/org/eclipse/xtext/ui/tests/core/resource/Storage2UriMapperJavaImplTest.java

```diff
@@ -0,0 +1,253 @@
+		JavaProjectSetupUtil.addJarToClasspath(project, file);
+		Assert.assertEquals("" + getCachedPackageFragmentRootData(), sizeBefore + 1, getCachedPackageFragmentRootData().size());
+		Assert.assertNotNull(IterableExtensions.findFirst(getCachedPackageFragmentRootData().keySet(), it -> it.contains("foo.jar")));
+		getCachedPackageFragmentRootData().entrySet().forEach((Map.Entry<String, PackageFragmentRootData> it) -> {
+			Assert.assertTrue(it.getValue().associatedRoots.size() + " / " + it.getKey(),
+					isExpectedJRESize(it.getValue().associatedRoots.size()));
+			String head = Iterables.getFirst(it.getValue().associatedRoots.keySet(), null);
+			Assert.assertTrue(head, head.startsWith("=testProject/"));
+		});
+		IJavaProject project2 = JavaProjectSetupUtil.createJavaProject("testProject2");
+		JavaProjectSetupUtil.addJarToClasspath(project2, file);
+		Assert.assertEquals("" + getCachedPackageFragmentRootData(), (sizeBefore + 1), getCachedPackageFragmentRootData().size());
+		Assert.assertNotNull(IterableExtensions.findFirst(getCachedPackageFragmentRootData().keySet(), it -> it.contains("foo.jar")));
+		getCachedPackageFragmentRootData().entrySet().forEach((Map.Entry<String, PackageFragmentRootData> it) -> {
+			Assert.assertTrue(it.getValue().associatedRoots.size() + " / " + it.getKey(),
+					isExpectedJRESize(it.getValue().associatedRoots.size(), 2));
+			String msg = Joiner.on("\n").join(it.getValue().associatedRoots.keySet());
+			Assert.assertTrue(msg, Iterables.any(it.getValue().associatedRoots.keySet(), r -> r.startsWith("=testProject/")));
+			Assert.assertTrue(msg, Iterables.any(it.getValue().associatedRoots.keySet(), r -> r.startsWith("=testProject2/")));
+		});
+		JavaProjectSetupUtil.removeJarFromClasspath(project, file);
+		Assert.assertEquals("" + getCachedPackageFragmentRootData(), sizeBefore + 1, getCachedPackageFragmentRootData().size());
+		Assert.assertNotNull(IterableExtensions.findFirst(getCachedPackageFragmentRootData().keySet(), it -> it.contains("foo.jar")));
+		getCachedPackageFragmentRootData().entrySet().forEach((Map.Entry<String, PackageFragmentRootData> it) -> {
+			if (it.getKey().contains("foo.jar")) {
+				Assert.assertEquals(it.getKey(), 1, it.getValue().associatedRoots.size());
+				String head = Iterables.getFirst(it.getValue().associatedRoots.keySet(), null);
+				Assert.assertTrue(head, head.startsWith("=testProject2/"));
+			} else {
+				Assert.assertTrue(it.getValue().associatedRoots.size() + "/" + it.getKey(),
+						isExpectedJRESize(it.getValue().associatedRoots.size(), 2));
+				String msg = IterableExtensions.join(it.getValue().associatedRoots.keySet(), "\n");
+				Assert.assertTrue(msg, Iterables.any(it.getValue().associatedRoots.keySet(), r -> r.startsWith("=testProject/")));
+				Assert.assertTrue(msg, Iterables.any(it.getValue().associatedRoots.keySet(), r -> r.startsWith("=testProject2/")));
+			}
+		});
+		JavaProjectSetupUtil.removeJarFromClasspath(project2, file);
```

Comment:
Would be nice if there could be more reuse between these blocks but not sure it's possible